### PR TITLE
[Core] Add Voxel Mesh Generation Modeler

### PR DESCRIPTION
**Pull request by jcotela**
**📝 Description**

This PR adds the Voxel Mesh Generator Modeler developed at @KratosMultiphysics/altair to the Kratos core. The modeler uses a modular approach to combine different strategies for its various steps:
- Generation of the key planes of the underlying regular grid
- Coloring of sets of voxels within the grid based on different criteria
- Mesh generation (which acts on all voxels of the same color) for tetrahedra, quadrilateral and their faces.
- ModelPart operations, which are post-process steps that can be applied after meshing and make use of the meshing data (currently used to find contacts between different sets of generated entities)

The modeler is registered in KratosCore so that it can be used through the standard factory.

Also included in the PR are C++ and Python tests for multiple meshing configurations.

**🆕 Changelog**
- Add VoxelMeshGeneratorModeler

### CODE REVIEW

File: kratos/tests/cpp_tests/modeler/test_key_plane_generation_with_refinement.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    model_part_io.ReadModelPart(model_part_0);
+
+    // Create the output .mdpa file
+    std::string output_file_name = "cube_skin_mesh";
+    std::fstream output_file;
+    output_file.open(output_file_name + ".mdpa", std::fstream::out);
+    output_file.close();
+
+    // Fill the output .mdpa file
+    ModelPartIO* model_part_io_write = new ModelPartIO(output_file_name, IO::WRITE);
```
Comment (philbucher) [inline kratos/tests/cpp_tests/modeler/test_key_plane_generation_with_refinement.cpp]: this leaks :D

File: kratos/tests/cpp_tests/modeler/test_key_plane_generation_with_refinement.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    model_part_io_write->WriteModelPart(model_part_0);
+}
+} // namespace
+
+KRATOS_TEST_CASE_IN_SUITE(VoxelMeshGeneratorModelerWithRefinement, KratosCoreFastSuite)
+{
+    using namespace Kratos;
+
+	WriteCubeSkinMeshMdpaFile();
+    std::cout << "Helloooo";
```
Comment (philbucher) [inline kratos/tests/cpp_tests/modeler/test_key_plane_generation_with_refinement.cpp]: :wave:

File: kratos/modeler/coloring/color_cells_with_inside_center.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                CheckGeometryType(r_geometry.GetGeometryType());
+                r_colors.AddGeometry(r_geometry, false);
+            }
+        } else if(input_entities == "geometries"){
+            for(auto& r_geometry : r_model_part.Geometries()) {
+                CheckGeometryType(r_geometry.GetGeometryType());
+                r_colors.AddGeometry(r_geometry, false);
+            }
+        } else{
+            KRATOS_ERROR << "The input_entities  " << parameters["input_entities"] << " is not supported. The supported input_entities are  elements and conditions" << std::endl;
```
Comment (philbucher) [inline kratos/modeler/coloring/color_cells_with_inside_center.h]: missing geometries

File: kratos/modeler/coloring/color_cells_with_inside_center.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                Element::GeometryType& r_geometry = r_element.GetGeometry();
+                CheckGeometryType(r_geometry.GetGeometryType());
+                r_colors.AddGeometry(r_geometry, false);
+            }
+        } else if(input_entities == "conditions"){
+            for(auto& r_condition : r_model_part.Conditions()) {
+                Condition::GeometryType& r_geometry = r_condition.GetGeometry();
+                CheckGeometryType(r_geometry.GetGeometryType());
+                r_colors.AddGeometry(r_geometry, false);
+            }
```
Comment (philbucher) [inline kratos/modeler/coloring/color_cells_with_inside_center.h]: conceptual question, should those still be allowed?
The overall goal is to move away from using elements and conditions in the input and only use geometries

File: kratos/modeler/key_plane_generation/key_plane_generation_by_outer_shell.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        auto dx = parameters["voxel_sizes"].GetVector();
+        // Now we find the min and max value for the node coordinates
+        Vector min_v(3,std::numeric_limits<double>::max()), max_v(3, -std::numeric_limits<double>::max());
+        for(auto& node: GetInputModelPart().Nodes()){
+            double x = node.X(), y = node.Y(), z = node.Z();
+            min_v[0] = std::min(min_v[0], x); max_v[0] = std::max(max_v[0], x);
+            min_v[1] = std::min(min_v[1], y); max_v[1] = std::max(max_v[1], y);
+            min_v[2] = std::min(min_v[2], z); max_v[2] = std::max(max_v[2], z);
+        }
```
Comment (philbucher) [inline kratos/modeler/key_plane_generation/key_plane_generation_by_outer_shell.h:56]: this can be very easily parallelized with the parallel utils

File: kratos/modeler/voxel_mesh_generator_modeler.cpp [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+        std::string model_part_name = parameters["model_part_name"].GetString();
+        std::string type = parameters["type"].GetString();
+        KRATOS_INFO("Modeler") << "Generate " << type << " in " << model_part_name  << " model part" << std::endl;
+
+        p_generator->Generate();
+    }
+    Timer::Stop("EntityGeneration");
+}
+
+ModelPart& VoxelMeshGeneratorModeler::CreateAndGetModelPart(std::string const& FullName)
```
Comment (philbucher) [inline kratos/modeler/voxel_mesh_generator_modeler.cpp:214]: I am pretty sure the `Model` does the same nowadays
@sunethwarna can you confirm pls?

File: kratos/modeler/voxel_mesh_generator_modeler.cpp [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+            if(!p_current_model_part->HasSubModelPart(name)){
+                p_current_model_part->CreateSubModelPart(name);
+            }
+            p_current_model_part = &(p_current_model_part->GetSubModelPart(name));
+        }
+
+        return *p_current_model_part;
+}
+
+ModelPart& VoxelMeshGeneratorModeler::GetModelPart(std::string const& FullName)
```
Comment (philbucher) [inline kratos/modeler/voxel_mesh_generator_modeler.cpp]: same, the Model does exactly this

File: kratos/modeler/voxel_mesh_generator_modeler.h [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+            const std::size_t index = GetNodeIndex(I,J,K);
+            return mNodalData[index];
+        }
+
+        const array_1d<std::size_t, 3> &GetNumberOfDivisions()
+        {
+            return mNumberOfDivisions;
+        }
+
+        bool IsInitialized()
```
Comment (philbucher) [inline kratos/modeler/voxel_mesh_generator_modeler.h]: const?

File: kratos/modeler/voxel_mesh_generator_modeler.h [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+    const Internals::SkinIntersection& GetIntersections(int Color) const;
+
+    bool IntersectionsGenerated(int Color) const;
+
+    ///@}
+    ///@name Serializer
+    ///@{
+
+    friend class Serializer;
```
Comment (philbucher) [inline kratos/modeler/voxel_mesh_generator_modeler.h]: I dont see this used, is it?

File: kratos/sources/kratos_application.cpp [hunk 10]
```diff
@@ -294,5 +300,11 @@ void KratosApplication::RegisterKratosCore() {
 
     // Register ConstitutiveLaw BaseClass
     KRATOS_REGISTER_CONSTITUTIVE_LAW("ConstitutiveLaw", mConstitutiveLaw);
+
+    //Register Voxel Modeler modular components
+    RegisterVoxelMesherColoring();
+    RegisterVoxelMesherKeyPlaneGeneration();
+    RegisterVoxelMesherEntityGeneration();
+    RegisterVoxelMesherOperation();
```
Comment (philbucher) [inline kratos/sources/kratos_application.cpp:308]: Is this really necessary? Usually this is done generic, why does the VoxelMesher require different treatment?

**Review by philbucher — COMMENTED** [review]
Great addition!

This is pretty large for reviewing, some initial comments/observations

File: kratos/sources/kratos_application.cpp (hunk 10 above)
Comment (jcotela) [reply kratos/sources/kratos_application.cpp:308]: I'm not sure if I understand you correctly but the only special treatment the VoxelMesher gets in terms of the registry is that it uses its own factory class. This is due to historical reasons (it started as a monolithic class and we eventually made the different components modular. The json input format that was already in use at that point is not compatible with what the standard Kratos factory expects, so we made our own factory class to parse the existing json)

**Review by loumalouomega — COMMENTED** [review]
A lot of code could be moved to source files, I will take a look later

**Comment by loumalouomega** [comment]
I don't know how, but I broke CentOS after https://github.com/KratosMultiphysics/Kratos/pull/12297/commits/3be5c902a485960c97f84b4a15ee21cf1833656a after just changing the headers¿?

File: kratos/modeler/coloring/color_cells_with_inside_center.h (hunk 4 above)
Comment (jcotela) [reply kratos/modeler/coloring/color_cells_with_inside_center.h]: I have nothing to say about the longer term plan, but I can tell you that we typically use this with a "traditional" mdpa file as input, so supporting elements is important for us at the moment.

File: kratos/sources/kratos_application.cpp (hunk 10 above)
Comment (jcotela) [reply kratos/sources/kratos_application.cpp:308]: Just to clarify, this Factory is for internal use within the modeler (it is used to instantiate the requested components on demand). The modeler itself is registered using the same approach as for other C++ modelers, with the usual kratos factory class.

File: kratos/modeler/voxel_mesh_generator_modeler.cpp (hunk 7 above)
Comment (jcotela) [reply kratos/modeler/voxel_mesh_generator_modeler.cpp]: This function was not used, so I'm removing it.

File: kratos/modeler/key_plane_generation/key_plane_generation_by_outer_shell.h (hunk 5 above)
Comment (jcotela) [reply kratos/modeler/key_plane_generation/key_plane_generation_by_outer_shell.h:56]: Agreed, but that requires a reduction to a bounding box, which we don't have at the moment. I would rather do that on a separate PR, with proper tests, rather than make this one bigger. It would also be a good candidate for a custom reduction in the sense of https://github.com/KratosMultiphysics/Kratos/pull/12195, if we go ahead with that one.

File: kratos/modeler/voxel_mesh_generator_modeler.cpp (hunk 6 above)
Comment (jcotela) [reply kratos/modeler/voxel_mesh_generator_modeler.cpp:214]: I am using the model capabilities now, but I'm leaving a helper function to check for existence of the new model part (as in that case the model method logs a warning)

File: kratos/modeler/voxel_mesh_generator_modeler.h (hunk 9 above)
Comment (jcotela) [reply kratos/modeler/voxel_mesh_generator_modeler.h]: Removed.

File: kratos/modeler/coloring/color_cells_with_inside_center.h (hunk 3 above)
Comment (jcotela) [reply kratos/modeler/coloring/color_cells_with_inside_center.h]: updated the message

**Comment by jcotela** [comment]
I think that I found the issue with the Centos7 CI, it turns out that the way some array_1d were being initialized from zero_vector was triggering an error if BOOST_UBLAS_TYPE_CHECK is enabled (I believe that this is implied by the -Wignored-qualifiers compiler flag? That macro is not defined in normal FullDebug builds)

@loumalouomega please let the CI run for this change before merging master, to see if we fixed the issue.

File: kratos/modeler/internals/cartesian_mesh_colors.h [hunk 11]
```diff
@@ -0,0 +1,811 @@
...
+
+        template<typename TPointsContainerType>
+        void CalculateMinMaxCenterOfElementPositions(TPointsContainerType const& Points, array_1d< std::size_t, 3 >& MinNodePosition, array_1d< std::size_t, 3 >& MaxNodePosition){
+            if(Points.empty())
+                return;
+
+            Point min_point;
+            Point max_point;
+            max_point = *(Points.begin());
+            min_point = *(Points.begin());
```
Comment (roigcarlo) [inline kratos/modeler/internals/cartesian_mesh_colors.h]:
```suggestion
            Point min_point = *(Points.begin());
            Point max_point = *(Points.begin());
```

File: kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+            std::vector<double> new_i_partitions;
+            std::vector<double> new_i_partitions_voxel_sizes;
+            // We add the first ones to the new partitions (leave the first point untouched)
+            new_i_partitions.push_back(*r_i_partition.begin());
+            new_i_partitions_voxel_sizes.push_back(*r_i_partitions_voxel_sizes.begin());
+            // Now we loop from the second one to the end and merge the ones in the middle
+            keep_merging = false;
+            if(r_i_partition.size()>3){
+                // The first one and the last one cannot be removed or moved.
+                for (int i = 1;  i+1 < (int) r_i_partition.size()-1; ++i) { //This is to ensure we don't remove the last one nor the first one so i+1 is the end - 1
```
Comment (roigcarlo) [inline kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp]:
```suggestion
                for (int i = 1;  i+1 < static_cast<int>(r_i_partition.size()-1); ++i) { //This is to ensure we don't remove the last one nor the first one so i+1 is the end - 1
```

(or `std::size_t i` if not going to parallelize)

Same with all the dynamic cast in the file

File: kratos/modeler/entity_generation/generate_tetrahedral_elements_with_cell_color.cpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+    auto& r_prototype_element = KratosComponents<Element>::Get(
+        parameters["generated_entity"].GetString());
+
+    array_1d<std::size_t, 3> number_of_cells = GetNumberOfCells();
+
+    ModelPart::NodesContainerType new_nodes;
+    ModelPart::ElementsContainerType new_elements;
+
+    KRATOS_INFO("Modeler") << "        Generating elements ... " << std::endl;
```
Comment (roigcarlo) [inline kratos/modeler/entity_generation/generate_tetrahedral_elements_with_cell_color.cpp]:
```suggestion
    KRATOS_INFO("Modeler") << "\tGenerating elements ... " << std::endl;
```

File: kratos/modeler/internals/cartesian_mesh_colors.h [hunk 14]
```diff
@@ -0,0 +1,811 @@
...
+      std::vector<array_1d<double,6>> mElementalFaceColors;
+      DenseMatrix<Internals::CartesianRay<Element::GeometryType>> mXYRays;
+      DenseMatrix<Internals::CartesianRay<Element::GeometryType>> mXZRays;
+      DenseMatrix<Internals::CartesianRay<Element::GeometryType>> mYZRays;
+
+      inline std::size_t ElementIndex(std::size_t I, std::size_t J, std::size_t K) const {
+        return I + J * mElementCenterCoordinates[0].size() + K * mElementCenterCoordinates[1].size() * mElementCenterCoordinates[0].size();
+      }
+     public:
+      CartesianMeshColors(): mTolerance(1e-12), mMinPoint(0.00, 0.00, 0.00), mMaxPoint(0.00, 0.00, 0.00){}
```
Comment (roigcarlo) [inline kratos/modeler/internals/cartesian_mesh_colors.h:52]: Why double `.00` instead of `0.0`?

Happens in several places.

File: kratos/modeler/internals/cartesian_mesh_colors.h [hunk 15]
```diff
@@ -0,0 +1,811 @@
...
+    void SetCoordinates(std::vector<double> const& rNodalXCoordinates, std::vector<double> const& rNodalYCoordinates, std::vector<double> const& rNodalZCoordinates){
+
+        KRATOS_ERROR_IF((rNodalXCoordinates.size() < 2) || (rNodalYCoordinates.size() < 2) || (rNodalZCoordinates.size() < 2))
+            << "The coordinates should have at least two entries defining the bounding box." << std::endl;
+
+        mNodalCoordinates[0] = rNodalXCoordinates;
+        mNodalCoordinates[1] = rNodalYCoordinates;
+        mNodalCoordinates[2] = rNodalZCoordinates;
+
+        for(int i = 0 ; i < 3 ; i++){
```
Comment (roigcarlo) [inline kratos/modeler/internals/cartesian_mesh_colors.h]: I would use `std::size_t` to keep it consistent.

Same with the other loops in the file.

File: kratos/modeler/internals/cartesian_mesh_colors.h [hunk 16]
```diff
@@ -0,0 +1,811 @@
...
+        std::vector<double>* p_z_coordinates = &(mNodalCoordinates[2]);
+
+        if((EntititesToColor == "center_of_elements") || (EntititesToColor == "face_of_elements")){
+            p_x_coordinates = &(mElementCenterCoordinates[0]);
+            p_y_coordinates = &(mElementCenterCoordinates[1]);
+            p_z_coordinates = &(mElementCenterCoordinates[2]);
+        }
+
+
+        #pragma omp parallel for
```
Comment (roigcarlo) [inline kratos/modeler/internals/cartesian_mesh_colors.h]: I know the parallelism will go in another PR but if you want to leave this one here it should be done with `block_for_each` or commented.

Same for all in the file.

File: kratos/modeler/internals/cartesian_ray.h [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+                if((i_intersection.first - Coordinates[current_index]) > NearEnough){
+                    ResultingColors[current_index++] = (is_inside) ? InsideColor : OutsideColor;
+                } else if((i_intersection.first - Coordinates[current_index]) > -NearEnough){ // Considering very near to wall as inside
+                    ResultingColors[current_index++] =  InsideColor;
+                }
+                else{
+                    break;
+                }
+            }
+            is_inside = (is_inside) ? false:true;
```
Comment (roigcarlo) [inline kratos/modeler/internals/cartesian_ray.h]:
```suggestion
            is_inside = !is_inside
```

File: kratos/modeler/internals/cartesian_ray.h [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+        const std::array<double, 8> delta_u{Delta, Delta, 0.00, -Delta, -Delta, -Delta, 0.00, Delta};
+        const std::array<double, 8> delta_v{0.00, Delta, Delta, Delta, 0.00, -Delta, -Delta, -Delta};
+        const std::array<int, 4> axes{1,2,0,1};
+
+        const int i_u = axes[mDirection];
+        const int i_v = axes[mDirection+1];
+
+        std::size_t no_hit_cases = 0;
+
+        for(int i_ray = 0 ; i_ray < 8 ; i_ray++){
```
Comment (roigcarlo) [inline kratos/modeler/internals/cartesian_ray.h]: use `std::size_t` if not going to be parallelized

File: kratos/modeler/internals/skin_intersection.h [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+    {
+        const std::size_t num_nodes = rGeom.PointsNumber();
+        const auto& r_integration_points = rGeom.IntegrationPoints(Method);
+        const std::size_t num_gauss = r_integration_points.size();
+
+        Matrix shape_functions = rGeom.ShapeFunctionsValues(Method);
+        Vector det_j;
+        rGeom.DeterminantOfJacobian(det_j, Method);
+
+        std::size_t offset = rIntegrationPointData.size();
```
Comment (roigcarlo) [inline kratos/modeler/internals/skin_intersection.h]: This is not used until the 2nd loop, move it with the other variables at the top or just before the 2nd loop.

File: kratos/modeler/key_plane_generation/key_plane_generation_by_bounding_box.cpp [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    Generate(parameters);
+}
+
+void KeyPlaneGenerationByBoundingBox::Generate(Parameters parameters) 
+{
+    array_1d<double,3> voxel_sizes = parameters["voxel_sizes"].GetVector();
+    array_1d<double,3> min_point = parameters["min_point"].GetVector();
+    array_1d<double,3> max_point = parameters["max_point"].GetVector();
+
+    for(int i_direction = 0 ; i_direction < 3 ; i_direction++){
```
Comment (roigcarlo) [inline kratos/modeler/key_plane_generation/key_plane_generation_by_bounding_box.cpp]: use `std::size_t` (Specially since `AddKeyPlane` expects that type)

File: kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+    }
+}
+
+std::pair<array_1d<std::vector<double>,3>,array_1d<std::vector<double>,3>> KeyPlaneGenerationWithRefinement::ComputePartitionsAndVoxelSizeByDirection(std::vector<BoundingBoxType>& rRefinementAreas,
+                                                std::vector<array_1d<double,3>>& rVoxelSizes,
+                                                BoundingBoxType& rGlobalBoundingBox,
+                                                const array_1d<double,3>& rGlobalVoxelSize){
+    array_1d<std::vector<double>,3> partitions;
+    array_1d<std::vector<double>,3> partitions_voxel_sizes;
+    for(int i_direction = 0 ; i_direction < 3 ; i_direction++){
```
Comment (roigcarlo) [inline kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp]: `std::size_t`. Happens on many other loops, to keep the review clean I will not add more.

File: kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+                                                int Direction,
+                                                double Position){
+    KRATOS_ERROR_IF(Position<rGlobalBoundingBox.first[Direction] || Position>rGlobalBoundingBox.second[Direction]) << "Trying to obtain the voxel size of a position outside the global bounding box." <<std::endl;
+    double dx = rGlobalVoxelSize[Direction];
+    for(int i = 0; i < (int)rRefinementAreas.size(); ++i){
+        auto& r_area = rRefinementAreas[i];
+        double& r_lbound = r_area.first[Direction];
+        double& r_ubound = r_area.second[Direction];
+        bool is_inside = (r_lbound<=Position && Position<=r_ubound);
+        if(is_inside)
```
Comment (roigcarlo) [inline kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp]:
```suggestion
        if (r_lbound<=Position && Position<=r_ubound)
```

File: kratos/modeler/operation/find_contacts_in_skin_model_part.cpp [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+    const CartesianMeshColors& rColors,
+    int& rPairingStatus,
+    double& rMinDistance,
+    int& rTemporaryColor) const
+{
+    constexpr double angle_threshold = 0.3;
+    int i = 0;
+    for(auto& ind: rNeighbourIndexes){
+        auto cell_center = rColors.GetCenterOfElement(ind[0], ind[1], ind[2]);
+        const int neigh_color = rNeighbourColors[i];
```
Comment (roigcarlo) [inline kratos/modeler/operation/find_contacts_in_skin_model_part.cpp]: Could use a `std::vector<int>::iterator rNeighbourColorsItr` and access to `*rNeighbourColorsItr` to avoid the auxiliary variable `i`

File: kratos/modeler/operation/find_contacts_in_skin_model_part.cpp [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+    const CartesianMeshColors& rColors,
+    std::vector<int>& rNeighbourColors,
+    std::vector<array_1d<std::size_t, 3>>& rNeighbourIndexes) const
+{
+    rNeighbourColors.clear();
+    rNeighbourIndexes.clear();
+    array_1d<int,6> i{-1,1,0,0,0,0};
+    array_1d<int,6> j{0,0,-1,1,0,0};
+    array_1d<int,6> k{0,0,0,0,-1,1};
+    for (int index = 0; index < 6; index++) {
```
Comment (roigcarlo) [inline kratos/modeler/operation/find_contacts_in_skin_model_part.cpp]: `index` appears marked in the highlighter as a reserved word/function. I am not aware of if its an intrinsic or smth, but maybe its a good idea to change.

File: kratos/modeler/operation/find_contacts_in_skin_model_part.cpp [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+        int i_tmp = rCellIndexes[0]+i[index];
+        int j_tmp = rCellIndexes[1]+j[index];
+        int k_tmp = rCellIndexes[2]+k[index];
+        if(IsElementInsideBounds(rColors, i_tmp, j_tmp, k_tmp)){
+            array_1d<std::size_t, 3> indexes{ (std::size_t)i_tmp, (std::size_t)j_tmp, (std::size_t)k_tmp};
```
Comment (roigcarlo) [inline kratos/modeler/operation/find_contacts_in_skin_model_part.cpp]: Static cast please.

File: kratos/modeler/voxel_mesh_generator_modeler.cpp [hunk 26]
```diff
@@ -0,0 +1,332 @@
...
+
+#include "voxel_mesh_generator_modeler.h"
+#include "coloring/voxel_mesher_coloring_factory.h"
+#include "key_plane_generation/key_plane_generation_factory.h"
+#include "entity_generation/voxel_mesher_entity_generation_factory.h"
+#include "operation/voxel_mesher_operation_factory.h"
+
+namespace Kratos
+{
+using SizeType = std::size_t;
```
Comment (roigcarlo) [inline kratos/modeler/voxel_mesh_generator_modeler.cpp]: std::size_t is being used directly in the file in most of the cases. Unless the implementation for some code is not dependent on the size I would just remove this.

File: kratos/modeler/voxel_mesh_generator_modeler.cpp [hunk 27]
```diff
@@ -0,0 +1,332 @@
...
+    if(data_file_name!=""){
+        ModelPartIO(data_file_name).ReadModelPart(*mpInputModelPart);
+    }
+
+    // Create the target model
+    const std::string output_model_part_name = mParameters["output_model_part_name"].GetString();
+    ModelPart& main_model_part = mpModel->HasModelPart(output_model_part_name)
+            ? mpModel->GetModelPart(output_model_part_name)
+            : mpModel->CreateModelPart(output_model_part_name);
+    return main_model_part;
```
Comment (roigcarlo) [inline kratos/modeler/voxel_mesh_generator_modeler.cpp]: Can return the ternary directly without assign to a ref.

File: kratos/modeler/voxel_mesh_generator_modeler.h [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of Modeler
+    KRATOS_CLASS_POINTER_DEFINITION(VoxelMeshGeneratorModeler);
+
+    typedef std::size_t SizeType;
+    typedef std::size_t IndexType;
```
Comment (roigcarlo) [inline kratos/modeler/voxel_mesh_generator_modeler.h]: Not used (that I've seen)

File: kratos/tests/cpp_tests/modeler/test_key_plane_generation_with_refinement.cpp [hunk 29]
```diff
@@ -0,0 +1,175 @@
...
+    ModelPartIO model_part_io_write(output_file_name, IO::WRITE);
+    model_part_io_write.WriteModelPart(model_part_0);
+}
+} // namespace
+
+KRATOS_TEST_CASE_IN_SUITE(VoxelMeshGeneratorModelerWithRefinement, KratosCoreFastSuite)
+{
+    using namespace Kratos;
+
+	WriteCubeSkinMeshMdpaFile();
```
Comment (roigcarlo) [inline kratos/tests/cpp_tests/modeler/test_key_plane_generation_with_refinement.cpp:128]: Please remove file after test has finished (for all the tests).

File: kratos/modeler/operation/find_contacts_in_skin_model_part.cpp [hunk 30]
```diff
@@ -0,0 +1,250 @@
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    if (pairing_status > 0) {
+        //We did find a temporary candidate , so we use this one
+        KRATOS_CRITICAL_SECTION
```
Comment (roigcarlo) [inline kratos/modeler/operation/find_contacts_in_skin_model_part.cpp:152]: I assume this is here because this code is called within a parallel section? (Or will be in the future)

**Review by roigcarlo — COMMENTED** [review]
Mostly pedantic comments from my part, and some remarks about `#pragma omp` in some places, but in general :+1:

File: kratos/modeler/voxel_mesh_generator_modeler.cpp (hunk 27 above)
Comment (jcotela) [reply kratos/modeler/voxel_mesh_generator_modeler.cpp]: Agreed, but I will change it to use an existing function that does this

File: kratos/modeler/entity_generation/generate_tetrahedral_elements_with_cell_color.cpp (hunk 13 above)
Comment (jcotela) [reply kratos/modeler/entity_generation/generate_tetrahedral_elements_with_cell_color.cpp]: Removing this printout (it is the only entity generator that prints to stdout)

File: kratos/modeler/operation/find_contacts_in_skin_model_part.cpp [hunk 31]
```diff
@@ -0,0 +1,250 @@
...
+        contact_container.mConditionsMap.insert(std::make_pair(outside_color, std::vector<std::size_t>()));
+        contact_container.mModelPartMap.insert(std::make_pair(outside_color, &r_model_part));
+        contact_container.mColorVector.push_back(outside_color);
+    }
+
+    ModelPart& r_skin_part = GetModelPart(parameters["model_part_name"].GetString());
+    const int cell_color = parameters["cell_color"].GetInt();
+    KRATOS_INFO("Modeler") << "Finding contacts for " << r_skin_part.FullName()  << " model part" << std::endl;
+
+    block_for_each(r_skin_part.Conditions(),[&](Condition& rCondition) {
```
Comment (loumalouomega) [inline kratos/modeler/operation/find_contacts_in_skin_model_part.cpp:67]: @roigcarlo @jcotela the critical is called here

File: kratos/modeler/operation/find_contacts_in_skin_model_part.cpp (hunk 30 above)
Comment (loumalouomega) [reply kratos/modeler/operation/find_contacts_in_skin_model_part.cpp:152]: I pointed where

**Comment by jcotela** [comment]
@roigcarlo I have removed the active pragma omp (replaced them by IndexPartitions) but I have left the ones that were commented out. I would like to keep them as a reminder of places were we disabled parallelism, since it might still be possible to make them parallel in the future. We will address those in a separate PR.

**Review by loumalouomega — CHANGES_REQUESTED** [review]
I think this is ready to goI

**Review by loumalouomega — APPROVED** [review]
